File: src/BOJ/python/BOJ_1149_RGB거리.py
# 1번 집의 색은 2번 집과 달라야 한다.
# N번 집의 색은 N - 1과 달라야 한다.
# i (2 ~ N - 1)의 집은 i - 1, i + 1과 달라야 한다.

import sys
input = sys.stdin.readline

N = int(input())
dp = [[0, 0, 0] for _ in range(N + 1)]

graph = [list(map(int, input().split())) for _ in range(N)]
graph.insert(0, [0, 0, 0])

dp[1] = graph[1][:]

for i in range(2, N + 1):
    dp[i][0] = min(dp[i - 1][1], dp[i - 1][2]) + graph[i][0]
    dp[i][1] = min(dp[i - 1][0], dp[i - 1][2]) + graph[i][1]
    dp[i][2] = min(dp[i - 1][0], dp[i - 1][1]) + graph[i][2]

print(dp[N])

File: src/BOJ/java/BOJ_1149_RGB거리.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class BOJ_1149_RGB거리 {
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());

        int[][] graph = new int[N + 1][];

        for (int i = 1; i <= N; i++) {
            st = new StringTokenizer(br.readLine());
            graph[i] = new int[] {
                    Integer.parseInt(st.nextToken()),
                    Integer.parseInt(st.nextToken()),
                    Integer.parseInt(st.nextToken())
            };
        }

        for (int i = 2; i <= N; i++) {
            graph[i][0] += Math.min(graph[i - 1][1], graph[i - 1][2]);
            graph[i][1] += Math.min(graph[i - 1][0], graph[i - 1][2]);
            graph[i][2] += Math.min(graph[i - 1][1], graph[i - 1][0]);
        }

        Arrays.sort(graph[N]);
        System.out.println(graph[N][0]);

    }
}
